app/pages/learn/01_tutorial/04_mastering-the-api/02_modern_io/index.md

```diff
@@ -0,0 +1,253 @@
+
+### Traversing Entries in Directories and Subdirectories
+
+For most situations you can use one of two methods. The `Files.list` method visits all entries (files, subdirectories, symbolic links) of a directory.
+
+```
+try (Stream<Path> entries = Files.list(pathToDirectory)) 
+{
+   . . .
+}
+```
+
+Use a `try`-with-resources statement to ensure that the stream object, which keeps track of the iteration, will be closed.
+
+If you also want to visit the entries of descendant directories, instead use the method
+
+```
+Stream<Path> entries = Files.walk(pathToDirectory);
+```
+
+Then simply use stream methods to home in on the entries that you are interested in, and to collect the results:
+
+```
+try (Stream<Path> entries = Files.walk(pathToDirectory)) {
+   List<Path> htmlFiles = entries.filter(p -> p.toString().endsWith("html")).toList();
+   . . .
+}
+```
+
+Here are the other methods for traversing directory entries:
+
+* An overloaded version of `Files.walk` lets you limit the depth of the traversed tree.
+* Two `Files.walkFileTree` methods provide more control over the iteration process, by notifying a `FileVisitor` when a directory is visited for the first and last time. This can be occasionally useful, in particularly for emptying and deleting a tree of directories. See the tutorial [Walking the File Tree](https://dev.java/learn/java-io/file-system/walking-tree) for details. Unless you need this control, use the simpler `Files.walk` method.
+* The `Files.find` method is just like `Files.walk`, but you provide a filter that inspects each path and its `BasicFileAttributes`. This is slightly more efficient than reading the attributes separately for each file.
+* Two `Files.newDirectoryStream` methods yields `DirectoryStream` instances, which can be used in enhanced `for` loops. There is no advantage over using `Files.list`. 
+* The legacy `File.list` or `File.listFiles` methods return file names or `File` objects. These are now obsolete.
+
+### Working with Zip Files
+
+Ever since Java 1.1, the `ZipInputStream` and `ZipOutputStream` classes provide an API for processing zip files. But the API is a bit clunky. Java 8 introduced a much nicer *zip file system*:
+
+```
+FileSystem fs = FileSystems.newFileSystem(pathToZipFile);
+```
+
+You can then use the methods of the `Files` class. Here we get a list of all files in the zip file:
+
+```
+try (Stream<Path> entries = Files.walk(fs.getPath("/"))) {
+   List<Path> filesInZip = entries.filter(Files::isRegularFile).toList();
+}
+```
+
+To read the file contents, just use `Files.readString` or `Files.readAllBytes`:
+
+```
+String contents = Files.readString(fs.getPath("/LICENSE"));
+```
+
+You can remove files with `Files.delete`. To add or replace files, simply use `Files.writeString` or `Files.write`.
+
+You must close the file system so that the changes are written to the zip file. Call
+
+```
+fs.close();
+```
+
+or use a `try`-with-resources statement.
```

Comment:
The `try`-with-resources approach is typically recommended in most situations.
What about changing the order to first write about `try`-with-resources and then mention the manual use of `close()` afterwards?

Reply:
I did that.